cobra/flux_analysis/deletion.py

```diff
@@ -0,0 +1,348 @@
+
+    Parameters
+    ----------
+    cobra_model : cobra.Model
+        The metabolic model to perform deletions in.
+
+    element_lists : list
+        List of lists `cobra.Reaction`s or `cobra.Gene`s (or their IDs)
+        to be deleted.
+
+    method: {"fba", "moma"}, optional
+        Procedure used to predict the growth rate.
+
+    solver: str, optional
+        This must be a QP-capable solver for MOMA. If left unspecified,
+        a suitable solver will be automatically chosen.
+
+    zero_cutoff: float, optional
+        When checking to see if a value is 0, this threshold is used.
+
+    Returns
+    -------
+    dict of dict
+        A sparse representation of all combinations of pairs of reaction
+        deletions without replacements.
+    """
+    with cobra_model as model:
+        WORKERS = dict(gene=_gene_deletion_worker, reaction=_reaction_deletion_worker)
+        worker_function = WORKERS[entity]
+
+        try:
+            (legacy, solver) = sutil.choose_solver(model, solver, qp=(method == "moma"))
+        except sutil.SolverNotFound:
+            if method == "moma":
+                warn("Cannot use MOMA since no QP-capable solver was found. Falling"
+                     " back to FBA.")
+                (legacy, solver) = sutil.choose_solver(model, solver)
+            else:
+                (err_type, err_val, err_tb) = sys.exc_info()
+                raise_(err_type, err_val, err_tb)  # reraise for Python2&3
+
+        if legacy:
+            raise ValueError(
+                "Legacy solvers are not supported any longer. Please use one of"
+                " the optlang solver interfaces instead.")
+
+        if num_cpu is None:
+            num_cpu = kwargs.get("number_of_processes")
+        if num_cpu is None:
+            try:
+                num_cpu = multiprocessing.cpu_count()
+            except NotImplementedError:
+                warn("Number of cores could not be detected - assuming 1.")
+                num_cpu = 1
+
+        # Reactions with 0 flux in the current conditions have no impact on
+        # deletion. Determine those reactions now to speed up computation later.
+        if method == 'moma':
+            moma.add_moma(model)
+        if method == 'linear moma':
+            moma.add_moma(model, linear=True)
+        solution = model.optimize()
+        if solution.status == "optimal":
+            kwargs["growth"] = solution.objective_value
+            kwargs["zero_flux_reactions"] = \
+                {rxn for (rxn, flux) in solution.fluxes.items()
+                 if abs(flux) < zero_cutoff}
+        else:
+            warn("Non-optimal ({0:s}) base solution.".format(solution.status))
+
+        def unordered_combination(reactions):
+            return frozenset([i.id for i in reactions if i.id not in kwargs["zero_flux_reactions"]])
+
+        args = set([unordered_combination(comb) for comb in product(*element_lists)])
```

Comment:
One strategy I rather liked in the old implementation was that not only are the `"zero_flux_reactions"` uninteresting for further computations but it's also clear that any double (multiple) knockout combination involving an essential reaction (from the single knockouts) will be essential. I don't know how much you gain from that in practice but it seems like the computation of a lot of combinations could be avoided that way.

Reply:
but to calculate essentiality, wouldn't you need to perform single knockouts for all the reactions and optimise? I think it'll bring more of extra optimisations to the process which can slow it down

Reply:
Yes, you would have to do that. Which is why I also suggested the more iterative approach in my review comment. I'm not sure the strategy is better. The problem is also that it's, of course, model dependent whether it's worth doing or not. Let *N* be the set of reactions with non-zero flux. Let *n* be the cardinality |*N*|. Let *K* be the set of essential reactions with *K* being a subset of *N*. Let *k* be the cardinality |*K*|. Let *C* be the binomial coefficient. Then it's worth doing this if *k* < *C*(*n*^2, 2) - *C*((*n* - *k*)^2, 2).